src/DependencyInjection/DI/src/ServiceLookup/CallSiteFactory.cs

```diff
@@ -129,54 +129,62 @@ private ServiceCallSite TryCreateOpenGeneric(Type serviceType, CallSiteChain cal
-                        var descriptor = descriptors[i];
+                        for (int i = 0; i < descriptors.Count; i++)
+                        {
+                            var descriptor = descriptors[i];
 
-                        // Last service should get slot 0
-                        var slot = descriptors.Count - i - 1;
-                        // There may not be any open generics here
-                        var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot);
-                        Debug.Assert(callSite != null);
+                            // Last service should get slot 0
+                            var slot = descriptors.Count - i - 1;
+                            // There may not be any open generics here
+                            var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot);
+                            Debug.Assert(callSite != null);
 
-                        callSites.Add(callSite);
+                            callSites.Add(callSite);
+                        }
                     }
-                }
-                else
-                {
-                    var slot = 0;
-                    // We are going in reverse so the last service in descriptor list gets slot 0
-                    for (var i = _descriptors.Count - 1; i >= 0; i--)
+                    else
                     {
-                        var descriptor = _descriptors[i];
-                        var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot) ??
-                                       TryCreateOpenGeneric(descriptor, itemType, callSiteChain, slot);
-                        slot++;
-                        if (callSite != null)
+                        var slot = 0;
+                        // We are going in reverse so the last service in descriptor list gets slot 0
+                        for (var i = _descriptors.Count - 1; i >= 0; i--)
                         {
-                            callSites.Add(callSite);
+                            var descriptor = _descriptors[i];
+                            var callSite = TryCreateExact(descriptor, itemType, callSiteChain, slot) ??
+                                           TryCreateOpenGeneric(descriptor, itemType, callSiteChain, slot);
+                            slot++;
+                            if (callSite != null)
+                            {
+                                callSites.Add(callSite);
+                            }
                         }
+
+                        callSites.Reverse();
                     }
 
-                    callSites.Reverse();
+                    return new IEnumerableCallSite(itemType, callSites.ToArray());
                 }
 
-                return new IEnumerableCallSite(itemType, callSites.ToArray());
+                return null;
+            }
+            finally
+            {
+                callSiteChain.Remove(serviceType);
```

Comment:
Is there still any reason to call remove in the CreateCallSite method?

Reply:
Good point.